File: skills/file_manager.py
import os
import logging
from pathlib import Path
from skills.base_skill import BaseSkill

logger = logging.getLogger(__name__)


class FileManagerSkill(BaseSkill):
    def __init__(self, allowed_directories: list[str] | None = None):
        self.allowed_directories = allowed_directories

    @property
    def name(self) -> str:
        return "file_manager"

    @property
    def description(self) -> str:
        return "Verwaltet Dateien: Lesen, Schreiben, Auflisten, Erstellen und Loeschen von Dateien und Ordnern."

    @property
    def parameters(self) -> dict:
        return {
            "type": "object",
            "properties": {
                "action": {
                    "type": "string",
                    "enum": ["read", "write", "list", "mkdir", "delete", "info"],
                    "description": "Die auszufuehrende Aktion",
                },
                "path": {
                    "type": "string",
                    "description": "Der Dateipfad",
                },
                "content": {
                    "type": "string",
                    "description": "Inhalt zum Schreiben (nur bei action=write)",
                },
            },
            "required": ["action", "path"],
        }

    def _check_access(self, path: str) -> bool:
        if self.allowed_directories is None:
            return True
        resolved = str(Path(path).resolve())
        return any(resolved.startswith(str(Path(d).resolve())) for d in self.allowed_directories)

    async def execute(self, action: str, path: str, content: str = "", **kwargs) -> str:
        path = os.path.expanduser(path)

        if not self._check_access(path):
            return f"Zugriff verweigert: '{path}' liegt nicht in den erlaubten Verzeichnissen."

        try:
            if action == "read":
                p = Path(path)
                if not p.exists():
                    return f"Datei nicht gefunden: {path}"
                text = p.read_text(encoding="utf-8", errors="replace")
                if len(text) > 10000:
                    text = text[:10000] + "\n... (gekuerzt)"
                return text

            elif action == "write":
                p = Path(path)
                p.parent.mkdir(parents=True, exist_ok=True)
                p.write_text(content, encoding="utf-8")
                return f"Datei geschrieben: {path}"

            elif action == "list":
                p = Path(path)
                if not p.exists():
                    return f"Verzeichnis nicht gefunden: {path}"
                entries = sorted(p.iterdir(), key=lambda x: (not x.is_dir(), x.name.lower()))
                lines = []
                for e in entries[:100]:
                    prefix = "[DIR]  " if e.is_dir() else "[FILE] "
                    size = ""
                    if e.is_file():
                        s = e.stat().st_size
                        size = f" ({s:,} bytes)"
                    lines.append(f"{prefix}{e.name}{size}")
                result = "\n".join(lines)
                if len(entries) > 100:
                    result += f"\n... und {len(entries) - 100} weitere Eintraege"
                return result or "(leeres Verzeichnis)"

            elif action == "mkdir":
                Path(path).mkdir(parents=True, exist_ok=True)
                return f"Verzeichnis erstellt: {path}"

            elif action == "delete":
                p = Path(path)
                if not p.exists():
                    return f"Nicht gefunden: {path}"
                if p.is_file():
                    p.unlink()
                    return f"Datei geloescht: {path}"
                else:
                    import shutil
                    shutil.rmtree(path)
                    return f"Verzeichnis geloescht: {path}"

            elif action == "info":
                p = Path(path)
                if not p.exists():
                    return f"Nicht gefunden: {path}"
                stat = p.stat()
                return (
                    f"Pfad: {p.resolve()}\n"
                    f"Typ: {'Verzeichnis' if p.is_dir() else 'Datei'}\n"
                    f"Groesse: {stat.st_size:,} bytes\n"
                    f"Geaendert: {stat.st_mtime}"
                )

            else:
                return f"Unbekannte Aktion: {action}"

        except Exception as e:
            logger.exception(f"file_manager {action} failed")
            return f"Fehler: {e}"

File: src/tools/builtin/file-manager.ts
import { readFileSync, writeFileSync, readdirSync, mkdirSync, rmSync, statSync, existsSync } from 'node:fs'
import path from 'node:path'
import os from 'node:os'
import type { Tool } from '../registry.js'

function expandPath(p: string): string {
  if (p.startsWith('~')) return path.join(os.homedir(), p.slice(1))
  return p
}

export class FileManagerTool implements Tool {
  readonly name = 'file_manager'
  readonly description = 'Verwaltet Dateien: Lesen, Schreiben, Auflisten, Erstellen und Loeschen von Dateien und Ordnern.'
  readonly parameters = {
    type: 'object',
    properties: {
      action: {
        type: 'string',
        enum: ['read', 'write', 'list', 'mkdir', 'delete', 'info'],
        description: 'Die auszufuehrende Aktion',
      },
      path: { type: 'string', description: 'Der Dateipfad' },
      content: { type: 'string', description: 'Inhalt zum Schreiben (nur bei action=write)' },
    },
    required: ['action', 'path'],
  }

  constructor(private readonly allowedDirectories: string[] | null = null) {}

  private _checkAccess(filePath: string): boolean {
    if (this.allowedDirectories === null) return true
    const resolved = path.resolve(filePath)
    return this.allowedDirectories.some(d => resolved.startsWith(path.resolve(d)))
  }

  async execute(args: Record<string, unknown>): Promise<string> {
    const action = String(args.action ?? '')
    const filePath = expandPath(String(args.path ?? ''))
    const content = String(args.content ?? '')

    if (!this._checkAccess(filePath)) {
      return `Zugriff verweigert: '${filePath}' liegt nicht in den erlaubten Verzeichnissen.`
    }

    try {
      switch (action) {
        case 'read': {
          if (!existsSync(filePath)) return `Datei nicht gefunden: ${filePath}`
          let text = readFileSync(filePath, 'utf-8')
          if (text.length > 10000) text = text.slice(0, 10000) + '\n... (gekuerzt)'
          return text
        }
        case 'write': {
          mkdirSync(path.dirname(filePath), { recursive: true })
          writeFileSync(filePath, content, 'utf-8')
          return `Datei geschrieben: ${filePath}`
        }
        case 'list': {
          if (!existsSync(filePath)) return `Verzeichnis nicht gefunden: ${filePath}`
          const entries = readdirSync(filePath, { withFileTypes: true })
          const sorted = entries.sort((a, b) => {
            if (a.isDirectory() !== b.isDirectory()) return a.isDirectory() ? -1 : 1
            return a.name.localeCompare(b.name)
          })
          const lines = sorted.slice(0, 100).map(e => {
            const prefix = e.isDirectory() ? '[DIR]  ' : '[FILE] '
            let size = ''
            if (e.isFile()) {
              try {
                const s = statSync(path.join(filePath, e.name)).size
                size = s < 1024 ? ` (${s} B)` : s < 1048576 ? ` (${(s / 1024).toFixed(1)} KB)` : ` (${(s / 1048576).toFixed(1)} MB)`
              } catch { /* skip */ }
            }
            return `${prefix}${e.name}${size}`
          })
          if (sorted.length > 100) lines.push(`... und ${sorted.length - 100} weitere Eintraege`)
          return lines.join('\n') || '(leer)'
        }
        case 'mkdir': {
          mkdirSync(filePath, { recursive: true })
          return `Verzeichnis erstellt: ${filePath}`
        }
        case 'delete': {
          if (!existsSync(filePath)) return `Nicht gefunden: ${filePath}`
          rmSync(filePath, { recursive: true, force: true })
          return `Geloescht: ${filePath}`
        }
        case 'info': {
          if (!existsSync(filePath)) return `Nicht gefunden: ${filePath}`
          const stat = statSync(filePath)
          return [
            `Pfad: ${filePath}`,
            `Typ: ${stat.isDirectory() ? 'Verzeichnis' : 'Datei'}`,
            `Groesse: ${stat.size} Bytes`,
            `Erstellt: ${stat.birthtime.toISOString()}`,
            `Geaendert: ${stat.mtime.toISOString()}`,
          ].join('\n')
        }
        default:
          return `Unbekannte Aktion: ${action}`
      }
    } catch (err) {
      return `Fehler: ${err instanceof Error ? err.message : String(err)}`
    }
  }
}
